itou/communications/models.py

```diff
@@ -0,0 +1,89 @@
+from django.conf import settings
+from django.contrib.contenttypes.fields import GenericForeignKey
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+
+
+class NotificationQuerySet(models.QuerySet):
+    def actives(self):
+        return self.exclude(is_obsolete=True)
+
+
+class Notification(models.Model):
+    notification_class = models.CharField(max_length=255, unique=True)
+    name = models.CharField(max_length=255)
+    category = models.CharField(max_length=255)
+    can_be_disabled = models.BooleanField(default=True)
+    is_obsolete = models.BooleanField(default=False, db_index=True)
+
+    objects = NotificationQuerySet.as_manager()
+
+    def __str__(self):
+        return self.name
+
+
+class DisabledNotification(models.Model):
+    notification = models.ForeignKey("Notification", on_delete=models.CASCADE)
+    settings = models.ForeignKey("NotificationSettings", on_delete=models.CASCADE)
+    disabled_at = models.DateTimeField(auto_now_add=True)
+
+    class Meta:
+        constraints = [
+            models.UniqueConstraint("notification", "settings", name="unique_notification_per_settings"),
+        ]
+
+
+class NotificationSettingsQuerySet(models.QuerySet):
+    def for_structure(self, structure=None):
+        qs = self.prefetch_related("disabled_notifications")
+        if structure is None:
+            return qs.filter(structure_type=None, structure_pk=None)
+        return qs.filter(structure_type=ContentType.objects.get_for_model(structure), structure_pk=structure.pk)
+
+
+class NotificationSettings(models.Model):
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, related_name="notification_settings")
+    structure_type = models.ForeignKey(ContentType, null=True, on_delete=models.CASCADE)
+    structure_pk = models.PositiveIntegerField(null=True)
+    structure = GenericForeignKey("structure_type", "structure_pk")
```

Comment:
En fait si on fait juste une `GenericForeignKey` alors Django ne semble pas gérer la suppression :
> If the related object is deleted, the `content_type` and `object_id` fields remain set to their original values and the `GenericForeignKey` returns `None`
> https://docs.djangoproject.com/en/5.0/ref/contrib/contenttypes/#django.contrib.contenttypes.fields.GenericForeignKey

Par contre ça le fait si on définis une `GenericRelation` de l'autre coté :
> Note also, that if you delete an object that has a `GenericRelation`, any objects which have a `GenericForeignKey` pointing at it will be deleted as well
> https://docs.djangoproject.com/en/5.0/ref/contrib/contenttypes/#django.contrib.contenttypes.fields.GenericRelation